core/src/main/java/com/linecorp/armeria/server/ContextPathServicesBuilder.java

```diff
@@ -16,45 +16,86 @@
- * built by this builder will be served under a context path.
+ * Builds {@link ServiceConfig}s for a {@link ServerBuilder} or {@link VirtualHostBuilder}.
+ * All {@link ServiceConfig}s built by this builder will be served under the specified context paths.
+ *
+ * <pre>{@code
+ * Server.builder()
+ *       .contextPath("/v1", "/v2")
+ *       .service(myService) // served under "/v1" and "/v2"
+ *       .and()
+ *       .virtualHost("foo.com")
+ *       .contextPath("/v3")
+ *       .service(myService) // served by virtual host "foo.com" under "/v3"
+ * }</pre>
  *
  * @param <T> the original type which will be returned once the {@link ServiceConfig}
  *            is built using {@link #and()}.
  */
-final class ContextPathServicesBuilder<T extends ServiceConfigsBuilder>
+@UnstableApi
+public final class ContextPathServicesBuilder<T extends ServiceConfigsBuilder>
         implements ServiceConfigsBuilder {
 
+    private final Set<String> contextPaths;
     private final T parent;
     private final VirtualHostBuilder virtualHostBuilder;
 
-    ContextPathServicesBuilder(T parent, VirtualHostBuilder virtualHostBuilder) {
+    ContextPathServicesBuilder(T parent, VirtualHostBuilder virtualHostBuilder,
+                               String... contextPaths) {
+        checkArgument(contextPaths.length > 0, "At least one context path is required");
+        for (String contextPath: contextPaths) {
+            RouteUtil.ensureAbsolutePath(contextPath, "contextPath");
+        }
+
         this.parent = parent;
+        this.contextPaths = ImmutableSet.copyOf(contextPaths);
         this.virtualHostBuilder = virtualHostBuilder;
     }
 
+    /**
+     * Configures an {@link HttpService} under the context path with the {@code customizer}.
+     */
+    public ContextPathServicesBuilder<T> withRoute(
+            Consumer<? super ContextPathServiceBindingBuilder<T>> customizer) {
+        customizer.accept(new ContextPathServiceBindingBuilder<>(this, contextPaths));
+        return this;
+    }
+
+    /**
+     * Returns a {@link ContextPathServiceBindingBuilder} which is for binding an {@link HttpService} fluently.
+     */
     @Override
-    public AbstractServiceBindingBuilder route() {
-        throw new UnsupportedOperationException("Not supported yet");
+    public ContextPathServiceBindingBuilder<T> route() {
+        return new ContextPathServiceBindingBuilder<>(this, contextPaths);
     }
 
+    /**
+     * Returns a {@link ContextPathDecoratingBindingBuilder} which is for binding a {@code decorator} fluently.
+     * The specified decorator(s) is/are executed in reverse order of the insertion.
+     */
     @Override
-    public AbstractBindingBuilder routeDecorator() {
-        throw new UnsupportedOperationException("Not supported yet");
+    public ContextPathDecoratingBindingBuilder<T> routeDecorator() {
+        return new ContextPathDecoratingBindingBuilder<>(this, contextPaths);
```

Comment:
`contextPaths()` could be exposed as a public API in `ContextPathServicesBuilder`. Afterward, an explicit constructor parameter is replaced with the getter.
```java
return new ContextPathServiceBindingBuilder<>(this);
```